pkg/controllers/nodeclaim/podevents/controller.go

```diff
@@ -0,0 +1,134 @@
+	// or if this is a daemonset
+	if pod.Spec.NodeName == "" || podutils.IsOwnedByDaemonSet(pod) {
+		return reconcile.Result{}, nil
+	}
+
+	node := &corev1.Node{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: pod.Spec.NodeName}, node); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("getting node, %w", err))
+	}
+
+	// If the node isn't owned by Karpenter, don't do anything
+	if _, ok := node.Labels[v1.NodePoolLabelKey]; !ok {
+		return reconcile.Result{}, nil
+	}
+
+	// If there's no associated node claim, it's not a karpenter owned node.
+	nodeClaims, err := nodeutils.GetNodeClaims(ctx, node, c.kubeClient)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(fmt.Errorf("getting nodeclaims for node, %w", err))
+	}
+
+	if len(nodeClaims) != 1 {
+		return reconcile.Result{}, fmt.Errorf("node %s did not have singular nodeclaim representation", node.Name)
+	}
+
+	nc := nodeClaims[0]
+	stored := nc.DeepCopy()
+
+	// If we've set the lastPodEvent before and it hasn't been before the timeout, don't do anything
+	if !nc.Status.LastPodEvent.Time.IsZero() && c.clock.Since(nc.Status.LastPodEvent.Time) < dedupeTimeout {
+		return reconcile.Result{}, nil
+	} else {
+		// otherwise, set the pod event time to now
+		nc.Status.LastPodEvent.Time = c.clock.Now()
+	}
+	if !equality.Semantic.DeepEqual(stored, nc) {
+		if err := c.kubeClient.Status().Patch(ctx, nc, client.MergeFrom(stored)); err != nil {
+			if errors.IsConflict(err) {
+				return reconcile.Result{Requeue: true}, nil
+			}
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named("nodeclaim.podevents").
+		For(&corev1.Pod{}).
+		WithEventFilter(predicate.TypedFuncs[client.Object]{
+			// If a pod is bound to a node or goes terminal
+			UpdateFunc: func(e event.TypedUpdateEvent[client.Object]) bool {
+				oldPod := (e.ObjectOld).(*corev1.Pod)
+				newPod := (e.ObjectNew).(*corev1.Pod)
+				// if this is a newly bound pod
+				bound := oldPod.Spec.NodeName == "" && newPod.Spec.NodeName != ""
+				// if this is a newly terminal pod
+				terminal := (newPod.Spec.NodeName != "" && !podutils.IsTerminal(oldPod) && podutils.IsTerminal(newPod))
+				// if this is a newly terminating pod
+				terminating := (newPod.Spec.NodeName != "" && !podutils.IsTerminating(oldPod) && podutils.IsTerminating(newPod))
+				// return true if it was bound to a node, went terminal, or went terminating
+				return bound || terminal || terminating
+			},
+			// If the pod is deleted, but wasn't already treated as terminal or terminating
+			DeleteFunc: func(e event.TypedDeleteEvent[client.Object]) bool {
+				pod := (e.Object).(*corev1.Pod)
+				return pod.Spec.NodeName != "" && !podutils.IsTerminal(pod) && !podutils.IsTerminating(pod)
```

Comment:
Is this going to work? The second you try to grab the pod with controller-runtime, it's not going to find it because there is no finalizer around. I'm also wondering how well this is going to work only because we may miss events this  way since if the pod is gone by the time we do the reconciliation, then we won't pick-up the pod and we will just exit without updating the nodeclaim at all